File: CovidWatch/app/src/main/java/com/riskre/covidwatch/ble/BLEContactEvent.java
package com.riskre.covidwatch.ble;

import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;

import java.nio.ByteBuffer;
import java.util.UUID;

import static com.riskre.covidwatch.UUIDs.CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC;
import static com.riskre.covidwatch.UUIDs.CONTACT_EVENT_IDENTIFIER_DESCRIPTOR;
import static com.riskre.covidwatch.UUIDs.CONTACT_EVENT_SERVICE;

public class BLEContactEvent {

    private static final String TAG = BLEContactEvent.class.getSimpleName();

    /**
     * Return a configured {@link BluetoothGattService} instance for the
     * Contact Event Service
     *
     * @returns BluetoothGattService for the CEN Orchestration
     */
    public static BluetoothGattService createContactEventService() {

        // Contact Event Service
        BluetoothGattService service = new BluetoothGattService(
                CONTACT_EVENT_SERVICE,
                BluetoothGattService.SERVICE_TYPE_PRIMARY);

        // Contact Event Identifier Characteristic
        BluetoothGattCharacteristic contactEventCharacteristic = new BluetoothGattCharacteristic(
                CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC,
                BluetoothGattCharacteristic.PROPERTY_READ,
                BluetoothGattCharacteristic.PERMISSION_READ);

        // Descriptor for the CEN
        BluetoothGattDescriptor UUIDDescriptor = new BluetoothGattDescriptor(
                CONTACT_EVENT_IDENTIFIER_DESCRIPTOR,
                BluetoothGattDescriptor.PERMISSION_READ);


        contactEventCharacteristic.addDescriptor(UUIDDescriptor);
        service.addCharacteristic(contactEventCharacteristic);

        return service;
    }

    /**
     * Generates a new Contact Event number. Currently generates a random UUID
     * and returns the 16 bytes
     *
     * @return  byte[] containing the UUID in big endian
     */
    public static byte[] getNewContactEventNumber() {

            UUID uuid = UUID.randomUUID();

            ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
            bb.putLong(uuid.getMostSignificantBits());
            bb.putLong(uuid.getLeastSignificantBits());

            return bb.array();
    }
}

File: CovidWatch/app/src/main/java/org/covidwatch/android/ble/BLEContactEvent.kt
package org.covidwatch.android.ble

import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattService
import org.covidwatch.android.utils.UUIDs.CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC
import org.covidwatch.android.utils.UUIDs.CONTACT_EVENT_IDENTIFIER_DESCRIPTOR
import org.covidwatch.android.utils.UUIDs.CONTACT_EVENT_SERVICE
import java.nio.ByteBuffer
import java.util.*


object BLEContactEvent {
    private val TAG = BLEContactEvent::class.java.simpleName

    /**
     * Return a configured [BluetoothGattService] instance for the
     * Contact Event Service
     *
     * @returns BluetoothGattService for the CEN Orchestration
     */
    fun createContactEventService(): BluetoothGattService {

        // Contact Event Service
        val service = BluetoothGattService(
            CONTACT_EVENT_SERVICE,
            BluetoothGattService.SERVICE_TYPE_PRIMARY
        )

        // Contact Event Identifier Characteristic
        val contactEventCharacteristic = BluetoothGattCharacteristic(
            CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC,
            BluetoothGattCharacteristic.PROPERTY_READ,
            BluetoothGattCharacteristic.PERMISSION_READ
        )

        // Descriptor for the CEN
        val UUIDDescriptor = BluetoothGattDescriptor(
            CONTACT_EVENT_IDENTIFIER_DESCRIPTOR,
            BluetoothGattDescriptor.PERMISSION_READ
        )
        contactEventCharacteristic.addDescriptor(UUIDDescriptor)
        service.addCharacteristic(contactEventCharacteristic)
        return service
    }

    /**
     * Generates a new Contact Event number. Currently generates a random UUID
     * and returns the 16 bytes
     *
     * @return  byte[] containing the UUID in big endian
     */
    val newContactEventNumber: ByteArray
        get() {
            val uuid = UUID.randomUUID()
            val bb = ByteBuffer.wrap(ByteArray(16))
            bb.putLong(uuid.mostSignificantBits)
            bb.putLong(uuid.leastSignificantBits)
            return bb.array()
        }
}